qermit/zero_noise_extrapolation/zne.py

```diff
@@ -683,45 +720,122 @@ def digital_folding_task_gen(
+
+
+def merge_experiments_task_gen() -> MitTask:
+    """Generates task merging qubit pauli strings when they belong to the
+    same experiment.
+
+    :return: MitTask performing the merge.
+    :rtype: MitTask
+    """
+
+    def task(
+        obj,
+        qpo_list: List[QubitPauliOperator],
+        experiment_index_list: List[int]
+    ) -> Tuple[List[QubitPauliOperator]]:
+        """Merge a list of qubit pauli operators. Qubit pauli operators will
+        merged if they belong to the same experiment. The experiment each
+        qubit pauli operator belongs to is indicated by the index list.
+        Merging here means that the mean of the pauli string coefficients
+        is taken to be the coefficient of the qubit pauli string in the
+        new qubit pauli operator.
+
+        :param qpo_list: List of qubit pauli strings, some of which may belong
+            to the same experiment.
+        :type qpo_list: List[QubitPauliOperator]
+        :param experiment_index_list: Indexes indicating to which experiment
+            the qubit pauli strings belong to.
+        :type experiment_index_list: List[int]
+        :return: A list of merged qubit pauli operators.
+        :rtype: Tuple[List[QubitPauliOperator]]
+        """
+
+        # A dictionary mapping the experiment index to the dictionary
+        # of the merged qubit pauli operator.
+        index_to_merged_qpo_dict: Dict[int, Dict[QubitPauliString, CoeffTypeAccepted]] = {}
+
+        # For each qubit pauli operator, sum the coefficients of the
+        # qubit pauli strings it contains.
+        for index, qpo in zip(experiment_index_list, qpo_list):
+
+            index_qpo_dict = index_to_merged_qpo_dict.get(index, {qps: 0 for qps in qpo._dict.keys()})
+
+            if not index_qpo_dict.keys() == qpo._dict.keys():
+                raise Exception(
+                    "The qubit pauli strings being merged do not contain "
+                    + "matching qubit pauli strings. In particular "
+                    + f"{index_qpo_dict.keys()} and {qpo._dict.keys()} differ."
+                )
+
+            # Sum qubit pauli string coefficients.
+            index_to_merged_qpo_dict[index] = {
+                qps: qpo._dict[qps] + index_qpo_dict[qps]
+                for qps in qpo._dict.keys()
+            }
+
+        # For each index, find the mean coefficient.
+        for index, merged_qpo_dict in index_to_merged_qpo_dict.items():
+            index_to_merged_qpo_dict[index] = {
+                qps: coeff / experiment_index_list.count(index)
+                for qps, coeff in merged_qpo_dict.items()
+            }
+
+        # Convert the dictionary to a list.
+        merged_qpo_list = [QubitPauliOperator({})] * (max(experiment_index_list) + 1)
+        for index in experiment_index_list:
+            merged_qpo_list[index] = QubitPauliOperator(dictionary=index_to_merged_qpo_dict[index])
+
+        return (merged_qpo_list, )
```

Comment:
You could try:
`return(merged_qpo_list = [
    QubitPauliOperator(index_to_merged_qpo_dict.get(index, {}))
    for index in range(max(experiment_index_list) + 1)
],)`

might be more confusing though?

If not I'd still type `merged_qpo_list: List[QubitPauliOperatpr]` - in this case I think it makes it clearer that this is the first time the object has been created, which is helpful to know at the bottom of a function